File: bindings/python/tests/test_runner.py
# ruff: noqa: D103,D100,S101

from __future__ import annotations

import openchecks
import pytest


def test_run_passed_success() -> None:
    class MockCheck(openchecks.BaseCheck):
        def check(self) -> openchecks.CheckResult[int]:
            return openchecks.CheckResult.passed("passed")

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

    check = MockCheck()
    result = openchecks.run(check)

    assert result.status() == openchecks.Status.Passed
    assert result.message() == "passed"


def test_run_failed_success() -> None:
    class MockCheck(openchecks.BaseCheck):
        def check(self) -> openchecks.CheckResult[int]:
            return openchecks.CheckResult.failed("failed")

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

    check = MockCheck()
    result = openchecks.run(check)

    assert result.status() == openchecks.Status.Failed
    assert result.message() == "failed"


def test_run_failed_check_returns_not_result() -> None:
    class MockCheck(openchecks.BaseCheck):
        def check(self) -> openchecks.CheckResult[int]:
            return None  # type: ignore

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

    check = MockCheck()
    result = openchecks.run(check)

    assert result.status() == openchecks.Status.SystemError


def test_run_failed_check_raises_error() -> None:
    exception = RuntimeError("test")

    class MockCheck(openchecks.BaseCheck):
        def check(self) -> openchecks.CheckResult[int]:
            raise exception

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

    check = MockCheck()
    result = openchecks.run(check)

    assert result.status() == openchecks.Status.SystemError
    err_result = result.error()
    assert err_result is not None
    assert isinstance(err_result, openchecks.CheckError)
    assert str(err_result) == str("RuntimeError: test")


def test_run_failed_check_does_not_inherit_base_check() -> None:
    check = None
    result = openchecks.run(check)  # type: ignore

    assert result.status() == openchecks.Status.SystemError


def test_auto_fix_passed_success() -> None:
    class MockCheck(openchecks.BaseCheck):
        def __init__(self) -> None:
            self._value = 1

        def check(self) -> openchecks.CheckResult[int]:
            if self._value != 0:
                return openchecks.CheckResult.failed(
                    "failed", [openchecks.Item(self._value)], can_fix=True
                )

            return openchecks.CheckResult.passed(
                "passed", [openchecks.Item(self._value)]
            )

        def auto_fix(self) -> None:
            self._value = 0

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

    check = MockCheck()
    result = openchecks.run(check)

    assert result.status() == openchecks.Status.Failed
    assert result.message() == "failed"

    result = openchecks.auto_fix(check)
    assert result.status() == openchecks.Status.Passed
    assert result.message() == "passed"


def test_auto_fix_failed_check_does_not_inherit_base_check() -> None:
    check = None
    result = openchecks.auto_fix(check)  # type: ignore

    assert result.status() == openchecks.Status.SystemError


def test_auto_fix_failed_check_hint_not_auto_fix() -> None:
    class MockCheck(openchecks.BaseCheck):
        def check(self) -> openchecks.CheckResult[int]:
            return openchecks.CheckResult.passed("passed")

        def auto_fix(self) -> None:
            pass

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

        def hint(self) -> openchecks.CheckHint:
            return openchecks.CheckHint.NONE

    check = MockCheck()
    result = openchecks.auto_fix(check)

    assert result.status() == openchecks.Status.SystemError


def test_auto_fix_failed_auto_fix_raises_error() -> None:
    exception = RuntimeError("Test")

    class MockCheck(openchecks.BaseCheck):
        def check(self) -> openchecks.CheckResult[int]:
            return openchecks.CheckResult.passed("passed")

        def auto_fix(self) -> None:
            raise exception

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

    check = MockCheck()
    result = openchecks.auto_fix(check)

    assert result.status() == openchecks.Status.SystemError

    err_result = result.error()
    assert err_result is not None
    assert isinstance(err_result, openchecks.CheckError)
    assert str(err_result) == "RuntimeError: Test"


@pytest.mark.asyncio()
async def test_async_run_passed_success() -> None:
    class MockCheck(openchecks.AsyncBaseCheck):
        async def async_check(self) -> openchecks.CheckResult[int]:
            return openchecks.CheckResult.passed("passed")

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

    check = MockCheck()
    result = await openchecks.async_run(check)

    assert result.status() == openchecks.Status.Passed
    assert result.message() == "passed"


@pytest.mark.asyncio()
async def test_async_run_failed_success() -> None:
    class MockCheck(openchecks.AsyncBaseCheck):
        async def async_check(self) -> openchecks.CheckResult[int]:
            return openchecks.CheckResult.failed("failed")

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

    check = MockCheck()
    result = await openchecks.async_run(check)

    assert result.status() == openchecks.Status.Failed
    assert result.message() == "failed"


@pytest.mark.asyncio()
async def test_async_run_failed_check_returns_not_result() -> None:
    class MockCheck(openchecks.AsyncBaseCheck):
        async def async_check(self) -> openchecks.CheckResult[int]:
            return None  # type: ignore

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

    check = MockCheck()
    result = await openchecks.async_run(check)

    assert result.status() == openchecks.Status.SystemError


@pytest.mark.asyncio()
async def test_async_run_failed_check_raises_error() -> None:
    exception = RuntimeError("test")

    class MockCheck(openchecks.AsyncBaseCheck):
        async def async_check(self) -> openchecks.CheckResult[int]:
            raise exception

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

    check = MockCheck()
    result = await openchecks.async_run(check)

    assert result.status() == openchecks.Status.SystemError
    err_result = result.error()
    assert err_result is not None
    assert isinstance(err_result, openchecks.CheckError)
    assert str(err_result) == "RuntimeError: test"


@pytest.mark.asyncio()
async def test_async_run_failed_check_does_not_inherit_base_check() -> None:
    check = None
    result = await openchecks.async_run(check)  # type: ignore

    assert result.status() == openchecks.Status.SystemError


@pytest.mark.asyncio()
async def test_async_auto_fix_passed_success() -> None:
    class MockCheck(openchecks.AsyncBaseCheck):
        def __init__(self) -> None:
            self._value = 1

        async def async_check(self) -> openchecks.CheckResult[int]:
            if self._value != 0:
                return openchecks.CheckResult.failed(
                    "failed", [openchecks.Item(self._value)], can_fix=True
                )

            return openchecks.CheckResult.passed(
                "passed", [openchecks.Item(self._value)]
            )

        async def async_auto_fix(self) -> None:
            self._value = 0

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

    check = MockCheck()
    result = await openchecks.async_run(check)

    assert result.status() == openchecks.Status.Failed
    assert result.message() == "failed"

    result = await openchecks.async_auto_fix(check)
    assert result.status() == openchecks.Status.Passed
    assert result.message() == "passed"


@pytest.mark.asyncio()
async def test_async_auto_fix_failed_check_does_not_inherit_base_check() -> None:
    check = None
    result = await openchecks.async_auto_fix(check)  # type: ignore

    assert result.status() == openchecks.Status.SystemError


@pytest.mark.asyncio()
async def test_async_auto_fix_failed_check_hint_not_auto_fix() -> None:
    class MockCheck(openchecks.AsyncBaseCheck):
        async def async_check(self) -> openchecks.CheckResult[int]:
            return openchecks.CheckResult.passed("passed")

        async def async_auto_fix(self) -> None:
            pass

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

        def hint(self) -> openchecks.CheckHint:
            return openchecks.CheckHint.NONE

    check = MockCheck()
    result = await openchecks.async_auto_fix(check)

    assert result.status() == openchecks.Status.SystemError


@pytest.mark.asyncio()
async def test_async_auto_fix_failed_auto_fix_raises_error() -> None:
    exception = RuntimeError("Test")

    class MockCheck(openchecks.AsyncBaseCheck):
        async def async_check(self) -> openchecks.CheckResult[int]:
            return openchecks.CheckResult.passed("passed")

        async def async_auto_fix(self) -> None:
            raise exception

        def title(self) -> str:
            return "title"

        def description(self) -> str:
            return "description"

    check = MockCheck()
    result = await openchecks.async_auto_fix(check)

    assert result.status() == openchecks.Status.SystemError

    err_result = result.error()
    assert err_result is not None
    assert isinstance(err_result, openchecks.CheckError)
    assert str(err_result) == "RuntimeError: Test"

File: bindings/cpp/tests/test_runner.cpp
#include <tuple>
#include <vector>

#include <gtest/gtest.h>

#include <openchecks/core.h>
#include <openchecks/item.h>
#include <openchecks/items.h>
#include <openchecks/result.h>
#include <openchecks/runner.h>
#include <openchecks/status.h>

#include "check_impl.h"
#include "item_impl.h"
#include "items_impl.h"
#include "result_impl.h"

class CheckParameterizedTestFixture
    : public ::testing::TestWithParam<
          std::tuple<std::string,                     // title
                     std::string,                     // description
                     OPENCHECKS_NAMESPACE::CheckHint, // hint
                     OPENCHECKS_NAMESPACE::Status,    // status
                     OPENCHECKS_NAMESPACE::Status,    // fix_status
                     std::string,                     // message
                     std::optional<IntItems>,         // items
                     bool,                            // can_fix
                     bool,                            // can_skip
                     std::optional<std::string>       // error
                     >> {};

TEST_P(CheckParameterizedTestFixture, ResultSuccess) {
  std::string title = std::get<0>(GetParam());
  std::string description = std::get<1>(GetParam());
  OPENCHECKS_NAMESPACE::CheckHint hint = std::get<2>(GetParam());
  OPENCHECKS_NAMESPACE::Status status = std::get<3>(GetParam());
  OPENCHECKS_NAMESPACE::Status fix_status = std::get<4>(GetParam());
  std::string message = std::get<5>(GetParam());
  std::optional<IntItems> items = std::get<6>(GetParam());
  bool can_fix = std::get<7>(GetParam());
  bool can_skip = std::get<8>(GetParam());
  std::optional<std::string> error = std::get<9>(GetParam());

  Check check = Check{title,   description, hint,    status,   fix_status,
                      message, items,       can_fix, can_skip, error};

  assert(check.title() == check._title);
  assert(check.description() == check._description);
  assert(check.hint() == check._hint);

  IntResult result = OPENCHECKS_NAMESPACE::run(check);

  OPENCHECKS_NAMESPACE::Status result_status = result.status();
  std::string_view result_message = result.message();
  std::optional<IntItems> result_items = result.items();
  std::optional<std::string_view> result_error = result.error();

  if (result_status == OPENCHECKS_NAMESPACE::Status::SystemError) {
    assert(result.can_fix() == false);
    assert(result.can_skip() == false);
  } else {
    assert(result.can_fix() == check._can_fix);
    assert(result.can_skip() == check._can_skip);
  }

  assert(result_message == check._message);
  assert(result_items == check._items);
  assert(result_error == check._error);

  if (result_status.has_failed() && result.can_fix()) {
    OPENCHECKS_NAMESPACE::CheckResult<int> fix_result =
        OPENCHECKS_NAMESPACE::auto_fix(check);

    OPENCHECKS_NAMESPACE::Status fix_result_status = fix_result.status();
    std::string_view fix_result_message = fix_result.message();
    const std::optional<IntItems> fix_result_items = fix_result.items();
    const std::optional<std::string_view> fix_result_error = fix_result.error();

    OPENCHECKS_NAMESPACE::CheckHint fix_hint = check.hint();

    if (!fix_hint.contains(OPENCHECKS_NAMESPACE::CheckHint::AutoFix)) {
      assert(fix_result_status == OPENCHECKS_NAMESPACE::Status::SystemError);
      assert(fix_result_message == "Check does not implement auto fix.");
      assert(fix_result_items == std::nullopt);
      assert(fix_result_error == std::nullopt);
    } else if (fix_result_error) {
      assert(fix_result_status == OPENCHECKS_NAMESPACE::Status::SystemError);
      assert(fix_result_message == "Error in auto fix.");
      assert(fix_result_items == std::nullopt);
      assert(fix_result_error.value() == check._error);
    } else {
      assert(fix_result_status == check._fix_status);
      assert(fix_result_message == check._message);

      if (!check._items) {
        assert(fix_result_items == std::nullopt);
      } else {
        assert(fix_result_items != std::nullopt);
        assert(fix_result_items == check._items);
      }

      assert(fix_result_error == std::nullopt);
    }

    if (fix_result_status == OPENCHECKS_NAMESPACE::Status::SystemError) {
      assert(fix_result.can_fix() == false);
      assert(fix_result.can_skip() == false);
    } else {
      assert(fix_result.can_fix() == check._can_fix);
      assert(fix_result.can_skip() == check._can_skip);
    }
  }
}

INSTANTIATE_TEST_SUITE_P(
    CheckResult, CheckParameterizedTestFixture,
    ::testing::Combine(
        ::testing::Values(std::string("title"), std::string("")), // title
        ::testing::Values(std::string("description"),
                          std::string("")), // description
        ::testing::Values(OPENCHECKS_NAMESPACE::CheckHint::None,
                          OPENCHECKS_NAMESPACE::CheckHint::AutoFix), // hint
        ::testing::Values(OPENCHECKS_NAMESPACE::Status::Pending,
                          OPENCHECKS_NAMESPACE::Status::Skipped,
                          OPENCHECKS_NAMESPACE::Status::Passed,
                          OPENCHECKS_NAMESPACE::Status::Warning,
                          OPENCHECKS_NAMESPACE::Status::Failed,
                          OPENCHECKS_NAMESPACE::Status::SystemError), // status
        ::testing::Values(
            OPENCHECKS_NAMESPACE::Status::Pending,
            OPENCHECKS_NAMESPACE::Status::Skipped,
            OPENCHECKS_NAMESPACE::Status::Passed,
            OPENCHECKS_NAMESPACE::Status::Warning,
            OPENCHECKS_NAMESPACE::Status::Failed,
            OPENCHECKS_NAMESPACE::Status::SystemError), // fix_status
        ::testing::Values(std::string("message"), std::string("")), // message
        ::testing::Values(IntItems{std::vector<IntItem>{
                              IntItem(0, ""), IntItem(1, ""), IntItem(2, "")}},
                          IntItems{std::vector<IntItem>{IntItem(0, "")}},
                          IntItems{std::vector<IntItem>{}},
                          std::nullopt), // items
        ::testing::Bool(),               // can_fix
        ::testing::Bool(),               // can_skip
        ::testing::Values(std::string("error"), std::string(),
                          std::nullopt) // error
        ));
